File: common/com.raytheon.uf.common.dataplugin.events.hazards/utility/common_static/base/python/events/HazardEvent.py
# #
# This software was developed and / or modified by Raytheon Company,
# pursuant to Contract DG133W-05-CQ-1067 with the US Government.
# 
# U.S. EXPORT CONTROLLED TECHNICAL DATA
# This software product contains export-restricted data whose
# export/transfer/disclosure is restricted by U.S. law. Dissemination
# to non-U.S. persons whether in the United States or abroad requires
# an export license or other authorization.
# 
# Contractor Name:        Raytheon Company
# Contractor Address:     6825 Pine Street, Suite 340
#                         Mail Stop B8
#                         Omaha, NE 68106
#                         402.291.0100
# 
# See the AWIPS II Master Rights File ("Master Rights File.pdf") for
# further licensing information.
# #


#
# A HazardEvent, based on Event, which returns information of hazardous activity
#  
#    
#     SOFTWARE HISTORY
#    
#    Date            Ticket#       Engineer       Description
#    ------------    ----------    -----------    --------------------------
#    01/22/13                      mnash          Initial Creation.
#    08/20/13        1360          blawrenc       Changed toStr() to __str__() for debugging
#    12/05/13        2527          bkowal         Removed obsolete conversion methods
# 
#

import JUtil, datetime
from shapely import wkt

from Event import Event

from java.util import Date
from com.raytheon.uf.common.dataplugin.events.hazards import HazardConstants, HazardConstants_ProductClass as ProductClass, HazardConstants_HazardStatus as HazardStatus


class HazardEvent(Event, JUtil.JavaWrapperClass):
        
    def __init__(self, wrappedObject):
        self.jobj = wrappedObject
        self.toPythonObj(wrappedObject)
        
    def getSiteID(self):
        return self.jobj.getSiteID()

    def setSiteID(self, site):
        self.jobj.setSiteID(site)
    
    def getEventID(self):
        return self.jobj.getEventID()
    
    def setEventID(self, eventId):
        self.jobj.setEventID(eventId)
    
    def getStatus(self):
        return self.getHazardStatus()
    
    def setStatus(self, hazardStatus):
        self.setHazardStatus(hazardStatus)

    def getHazardStatus(self):
        return self.jobj.getStatus().name()
    
    def setHazardStatus(self, hazardStatus):
        if hazardStatus is not None :
            self.jobj.setStatus(HazardStatus.valueOf(str(hazardStatus).upper()))
        else :
            self.jobj.setStatus(HazardStatus.PENDING)
            
    def getPhenomenon(self):
        return self.jobj.getPhenomenon()
    
    def setPhenomenon(self, phenomenon):
        self.jobj.setPhenomenon(phenomenon)
        
    def getSignificance(self):
        return self.jobj.getSignificance()
    
    def setSignificance(self, significance):
        self.jobj.setSignificance(significance)
        
    def getSubType(self):
        return self.jobj.getSubType()
    
    def setSubType(self, subtype):
        self.jobj.setSubType(subtype)
        
    def getHazardType(self):
        return self.jobj.getHazardType()
        
    def getCreationTime(self):
        return datetime.datetime.fromtimestamp(self.jobj.getCreationTime().getTime() / 1000.0)
    
    def setCreationTime(self, creationTime):
        dt = Date(long(self._getMillis(creationTime)))
        self.jobj.setCreationTime(dt)
      
    def getEndTime(self):
        if self.jobj.getEndTime() is not None :
            return datetime.datetime.fromtimestamp(self.jobj.getEndTime().getTime() / 1000.0)
        else :
            return None
          
    def setEndTime(self, endTime):
        dt = Date(long(self._getMillis(endTime)))
        self.jobj.setEndTime(dt)
    
    def getStartTime(self):
        return datetime.datetime.fromtimestamp(self.jobj.getStartTime().getTime() / 1000.0)
    
    def setStartTime(self, startTime):
        dt = Date(long(self._getMillis(startTime)))
        self.jobj.setStartTime(dt)
    
    def getGeometry(self):
        return JUtil.javaObjToPyVal(self.jobj.getGeometry())
    
    def setGeometry(self, geometry):
        if geometry is not None :
            self.jobj.setGeometry(JUtil.pyValToJavaObj(geometry))
    
    def getHazardMode(self):
        if self.jobj.getHazardMode() is not None :
            return self.jobj.getHazardMode().name()
        else :
            return None
    
    def setHazardMode(self, hazardMode):
        try :
            self.jobj.setHazardMode(HazardConstants.productClassFromAbbreviation(hazardMode))
        except Exception, e :
            try :
                self.jobj.setHazardMode(HazardConstants.productClassFromName(hazardMode))
            except Exception, e:
                self.jobj.setHazardMode(ProductClass.TEST)        
        
    def getHazardAttributes(self):
        return JUtil.javaObjToPyVal(self.jobj.getHazardAttributes())
    
    def setHazardAttributes(self, hazardAttributes):
        self.jobj.setHazardAttributes(JUtil.pyValToJavaObj(hazardAttributes))
        
    def addHazardAttribute(self, key, value):
        self.jobj.addHazardAttribute(key, JUtil.pyValToJavaObj(value))
        
    def removeHazardAttribute(self, key):
        self.jobj.removeHazardAttribute(key)

    def get(self, key, default=None):
        '''
         Get the value of the hazard attribute with given key.
         '''
        value = JUtil.javaObjToPyVal(self.jobj.getHazardAttribute(key))
        if not value:
            value = default
        return value
    
    def set(self, key, value):
        '''
        Set the hazard attribute with given key to given value.
        '''
        self.jobj.addHazardAttribute(key, JUtil.pyValToJavaObj(value))

    def addToList(self, key, value):
        '''
        The equivalent of hazardAttributes.setdefault(key, []).append(value)
        '''
        currentVal = JUtil.javaObjToPyVal(self.jobj.getHazardAttribute(key))
        if currentVal:
            currentVal.append(value)
        else:
            currentVal = [value]
        self.jobj.addHazardAttribute(key, JUtil.pyValToJavaObj(currentVal))
               
    def _getMillis(self, date):
        epoch = datetime.datetime.utcfromtimestamp(0)
        delta = date - epoch
        return delta.total_seconds() * 1000.0
    
    def toPythonObj(self, javaClass):
        '''
        @summary: Converts a Java HazardEvent to the corresponding Python version
        '''  
        self.setSiteID(javaClass.getSiteID())
        if javaClass.getStatus() is not None:    
            self.setHazardStatus(javaClass.getStatus().name())
        else :
            self.setHazardStatus(HazardStatus.PENDING)
        self.setPhenomenon(javaClass.getPhenomenon())
        self.setSignificance(javaClass.getSignificance())
        self.setSubType(javaClass.getSubType())
        if javaClass.getCreationTime() is not None:
            self.setCreationTime(datetime.datetime.fromtimestamp(javaClass.getCreationTime().getTime() / 1000.0))
        else :
            self.setCreationTime(datetime.datetime.now())
        if javaClass.getStartTime() is not None:
            self.setStartTime(datetime.datetime.fromtimestamp(javaClass.getStartTime().getTime() / 1000.0))
        else:
            self.setStartTime(datetime.datetime.now())
        if javaClass.getEndTime() is not None:
            self.setEndTime(datetime.datetime.fromtimestamp(javaClass.getEndTime().getTime() / 1000.0))
        else:
            self.setEndTime(datetime.datetime.now())
        if javaClass.getHazardMode() is not None:
            self.setHazardMode(javaClass.getHazardMode().name())
        else :
            self.setHazardMode(None)
        if javaClass.getGeometry() is not None :
            self.setGeometry(wkt.loads(javaClass.getGeometry().toText()))
        else :
            self.setGeometry(None)
        self.setHazardAttributes(JUtil.javaObjToPyVal(javaClass.getHazardAttributes()))
    
    def __getitem__(self, key):
        lowerKey = key.lower()
        if lowerKey == 'site':
            return getSiteID()
        elif lowerKey == 'status':
            return self.getStatus()
        elif lowerKey == 'phenomenon' or lowerKey == 'phen':
            return self.getPhenomenon()
        elif lowerKey == 'significance' or lowerKey == 'sig':
            return self.getSignificance()
        elif lowerKey == 'subtype':
            return self.getSubType()
        elif lowerKey == 'creationtime':
            return self.getCreationTime()
        elif lowerKey == 'endtime':
            return self.getEndTime()
        elif lowerKey == 'starttime': 
            return self.getStartTime()
        elif lowerKey == 'geometry' or lowerKey == 'geom':
            return self.getGeometry()
        elif lowerKey == 'mode' or lowerKey == 'hazardmode':
            return self.getHazardMode()
        elif lowerKey == 'eventid':
            return self.getEventID()
        elif lowerKey == 'attributes':
            return self.getHazardAttributes()
        else :
            raise TypeError()
        
    def __setitem__(self, key, value):
        lowerKey = key.lower()
        if lowerKey == 'site':
            self.setSiteID(value)
        elif lowerKey == 'status':
            self.setStatus(value)
        elif lowerKey == 'phenomenon' or lowerKey == 'phen':
            self.setPhenomenon(value)
        elif lowerKey == 'significance' or lowerKey == 'sig':
            self.setSignificance(value)
        elif lowerKey == 'subtype':
            self.setSubType(value)
        elif lowerKey == 'creationtime':
            self.setCreationTime(value)
        elif lowerKey == 'endtime':
            self.setEndTime(value)
        elif lowerKey == 'starttime': 
            self.setStartTime(value)
        elif lowerKey == 'geometry' or lowerKey == 'geom':
            self.setGeometry(value)
        elif lowerKey == 'mode' or lowerKey == 'hazardmode':
            self.setHazardMode(value)
        elif lowerKey == 'attributes':
            self.setHazardAttributes(value)         
    
    def __eq__(self, other):
        return self.jobj.equals(other.jobj)
    
    def __ne__(self, other):
        return not self.__eq__(other)
    
    def __str__(self):
        string = 'HazardEvent: ' + self.jobj.toString() + \
            '\ngeometry: ' + str(self.jobj.getGeometry())
        return string
    
    def toJavaObj(self):
        return self.jobj

File: common/com.raytheon.uf.common.dataplugin.events.hazards/src/com/raytheon/uf/common/dataplugin/events/hazards/event/HazardEvent.java
/**
 * This software was developed and / or modified by Raytheon Company,
 * pursuant to Contract DG133W-05-CQ-1067 with the US Government.
 * 
 * U.S. EXPORT CONTROLLED TECHNICAL DATA
 * This software product contains export-restricted data whose
 * export/transfer/disclosure is restricted by U.S. law. Dissemination
 * to non-U.S. persons whether in the United States or abroad requires
 * an export license or other authorization.
 * 
 * Contractor Name:        Raytheon Company
 * Contractor Address:     6825 Pine Street, Suite 340
 *                         Mail Stop B8
 *                         Omaha, NE 68106
 *                         402.291.0100
 * 
 * See the AWIPS II Master Rights File ("Master Rights File.pdf") for
 * further licensing information.
 **/
package com.raytheon.uf.common.dataplugin.events.hazards.event;

import java.io.Serializable;
import java.util.Date;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;
import java.util.UUID;

import javax.xml.bind.annotation.XmlAccessType;
import javax.xml.bind.annotation.XmlAccessorType;
import javax.xml.bind.annotation.XmlAttribute;
import javax.xml.bind.annotation.XmlElement;
import javax.xml.bind.annotation.XmlRootElement;
import javax.xml.bind.annotation.adapters.XmlJavaTypeAdapter;

import com.raytheon.uf.common.dataplugin.events.IValidator;
import com.raytheon.uf.common.dataplugin.events.ValidationException;
import com.raytheon.uf.common.dataplugin.events.hazards.HazardConstants;
import com.raytheon.uf.common.dataplugin.events.hazards.HazardConstants.HazardStatus;
import com.raytheon.uf.common.dataplugin.events.hazards.HazardConstants.ProductClass;
import com.raytheon.uf.common.dataplugin.events.hazards.registry.geometryadapters.GeometryAdapter;
import com.raytheon.uf.common.dataplugin.events.hazards.registry.slotconverter.HazardAttributeSlotConverter;
import com.raytheon.uf.common.message.WsId;
import com.raytheon.uf.common.message.adapter.WsIdAdapter;
import com.raytheon.uf.common.registry.annotations.RegistryObject;
import com.raytheon.uf.common.registry.annotations.RegistryObjectVersion;
import com.raytheon.uf.common.registry.annotations.SlotAttribute;
import com.raytheon.uf.common.registry.annotations.SlotAttributeConverter;
import com.raytheon.uf.common.registry.ebxml.slots.DateSlotConverter;
import com.raytheon.uf.common.serialization.annotations.DynamicSerialize;
import com.raytheon.uf.common.serialization.annotations.DynamicSerializeElement;
import com.vividsolutions.jts.geom.Geometry;

import gov.noaa.gsd.common.utilities.geometry.AdvancedGeometryAdapter;
import gov.noaa.gsd.common.utilities.geometry.AdvancedGeometryUtilities;
import gov.noaa.gsd.common.utilities.geometry.IAdvancedGeometry;
import gov.noaa.gsd.common.visuals.VisualFeature;
import gov.noaa.gsd.common.visuals.VisualFeaturesList;
import gov.noaa.gsd.common.visuals.VisualFeaturesListAdapter;

/**
 * The Hazard record class which at its most basic level contains information
 * that every hazard will contain
 * 
 * <pre>
 * 
 * SOFTWARE HISTORY
 * 
 * Date         Ticket#    Engineer    Description
 * ------------ ---------- ----------- --------------------------
 * Aug 16, 2012            mnash       Initial creation
 * Nov 04, 2013 2182     daniel.s.schaffer@noaa.gov      Started refactoring
 * Nov 14, 2013 1472       bkowal      Remove ISerializableObject. Renamed hazard subtype to subType.
 * Dec     2013 2368       thansen   Added getHazardType
 * Apr 23, 2014 2925       Chris.Golden Augmented with additional methods to
 *                                      set the type components atomically, or
 *                                      the start and end time atomically.
 * Jun 30, 2014 3512       Chris.Golden Added addHazardAttributes() method.
 * Feb 22, 2015 6561       mpduff      Override getInsertTime
 * May 29, 2015 6895      Ben.Phillippe Refactored Hazard Service data access
 * Jul 31, 2015 7458      Robert.Blum   Added new userName and workstation fields.
 * Aug 03, 2015 8836       Chris.Cody   Changes for a configurable Event Id
 * Mar 01, 2016 15676      Chris.Golden Added visual features to hazard event.
 * Mar 26, 2016 15676      Chris.Golden Added more methods to get and set
 *                                      individual visual features.
 * May 02, 2016 18235      Chris.Golden Added source field.
 * May 10, 2016 18515      Chris.Golden Fixed bug that caused duplicate-key hazard
 *                                      attributes to be kept around, so that a
 *                                      particular key (e.g. "cta") might have
 *                                      two or more entries in the attributes
 *                                      set, each with a different value. This
 *                                      same bug also caused attributes that were
 *                                      supposedly removed from the event to not
 *                                      actually be removed.
 * Jun 03, 2016 19403      mduff        Removed some slots.
 * Jun 08, 2016 18899     Ben.Phillippe Changed geometry fields to be xml elements
 *                                      instead of  attributes to avoid the jaxb
 *                                      attribute size limit
 * Jun 10, 2016 19537      Chris.Golden Combined base and selected visual feature
 *                                      lists for each hazard event into one,
 *                                      replaced by visibility constraints
 *                                      based upon selection state to individual
 *                                      visual features.
 * Jun 23, 2016 19537      Chris.Golden Changed to use new visual feature list
 *                                      method for visual feature replacement.
 * Sep 12, 2016 15934      Chris.Golden Changed hazard events to use advanced
 *                                      geometries instead of JTS geometries.
 * Sep 21, 2016 15934      Chris.Golden Changed to work with new version of
 *                                      AdvancedGeometryUtilities.
 * Dec 19, 2016 21504      Robert.Blum  Updates for hazard locking.
 * Feb 01, 2017 15556      Chris.Golden Added visible-in-history-list flag.
 * Feb 13, 2017 28892      Chris.Golden Removed slot converter for visual features
 *                                      list, as visual features should not be
 *                                      put in slots.
 * Feb 16, 2017 29138      Chris.Golden Removed the visible-in-history-list flag
 *                                      since use of the history list is being
 *                                      reduced with advent of ability to save
 *                                      a "latest version" to the database that
 *                                      is not part of the history list. Also
 *                                      added ability to configure instances to
 *                                      indicate they are "latest version" ones,
 *                                      that is, not to be in the history list.
 * Mar 30, 2017 15528      Chris.Golden Added modified flag as part of basic
 *                                      hazard event, since this flag must be
 *                                      persisted as part of the hazard event.
 * May 24, 2017 15561      Chris.Golden Added getPhensig() method.
 * Dec 17, 2017 20739      Chris.Golden Refactored away access to directly
 *                                      mutable session events.
 * </pre>
 * 
 * @author mnash
 * @version 1.0
 */

@XmlRootElement(name = "hazard")
@XmlAccessorType(XmlAccessType.NONE)
@DynamicSerialize
@RegistryObject({ HazardConstants.SITE_ID,
        HazardConstants.HAZARD_EVENT_IDENTIFIER, HazardConstants.UNIQUE_ID })
@RegistryObjectVersion(value = 1.0f)
public class HazardEvent implements IHazardEvent, IValidator {

    /**
     * Value for the {@link HazardEvent#uniqueID} indicating that this is the
     * latest version of the hazard event that has been persisted (not part of
     * the history list).
     */
    public static final String LATEST_VERSION = "latest";

    /**
     * Flag indicating whether or not the hazard event is in a modified state.
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.MODIFIED)
    private boolean modified;

    /**
     * The issuing site ID
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.SITE_ID)
    private String siteID;

    /**
     * The event ID
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.HAZARD_EVENT_IDENTIFIER)
    private String eventID;

    @DynamicSerializeElement
    @XmlAttribute
    private String uniqueID = UUID.randomUUID().toString();

    /**
     * The status of the record at this point in time
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.HAZARD_EVENT_STATUS)
    private HazardStatus status;

    /**
     * Phenomenon that is being recorded
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.PHENOMENON)
    private String phenomenon;

    /**
     * Significance of the hazard
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.SIGNIFICANCE)
    private String significance;

    /**
     * subtype of the hazard
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.HAZARD_EVENT_SUB_TYPE)
    private String subType;

    /**
     * The start time of the hazard
     */
    @DynamicSerializeElement
    @XmlElement
    @SlotAttribute(HazardConstants.HAZARD_EVENT_START_TIME)
    @SlotAttributeConverter(DateSlotConverter.class)
    private Date startTime;

    /**
     * The end time of the hazard
     */
    @DynamicSerializeElement
    @XmlElement
    @SlotAttribute(HazardConstants.HAZARD_EVENT_END_TIME)
    @SlotAttributeConverter(DateSlotConverter.class)
    private Date endTime;

    /**
     * The time this hazard was created
     */
    @DynamicSerializeElement
    @XmlElement
    private Date creationTime;

    /**
     * The mode of the hazard
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.HAZARD_MODE)
    private ProductClass hazardMode;

    /**
     * The source of the hazard
     */
    @DynamicSerializeElement
    @XmlAttribute
    @SlotAttribute(HazardConstants.HAZARD_SOURCE)
    private Source source;

    /**
     * The flattened geometry coverage of the hazard
     */
    @DynamicSerializeElement
    @XmlJavaTypeAdapter(value = GeometryAdapter.class)
    @XmlElement
    private Geometry flattenedGeometry;

    /**
     * The geometry coverage of the hazard.
     */
    @DynamicSerializeElement
    @XmlJavaTypeAdapter(value = AdvancedGeometryAdapter.class)
    @XmlElement
    private IAdvancedGeometry geometry;

    /**
     * Visual features list of the hazard.
     * <p>
     * No slot attribute or slot converter is provided because there is no
     * reason why visual features should have their own slots.
     * </p>
     */
    @DynamicSerializeElement
    @XmlJavaTypeAdapter(value = VisualFeaturesListAdapter.class)
    @XmlElement
    private VisualFeaturesList visualFeatures;

    /**
     * The time this hazard was inserted into the repository.
     */
    @DynamicSerializeElement
    @XmlElement
    @SlotAttribute(HazardConstants.INSERT_TIME)
    @SlotAttributeConverter(DateSlotConverter.class)
    private Date insertTime;

    /**
     * The workstation id of the person who created the hazard or the last
     * person that issue the hazard.
     */
    @DynamicSerializeElement
    @XmlElement
    @XmlJavaTypeAdapter(value = WsIdAdapter.class)
    private WsId wsId;

    /**
     * Additional attributes of the hazard.
     */
    @DynamicSerializeElement
    @XmlElement(name = "Attributes")
    @SlotAttribute("Attributes")
    @SlotAttributeConverter(HazardAttributeSlotConverter.class)
    private Set<HazardAttribute> attributes = new HashSet<HazardAttribute>();

    /**
     * Construct a standard instance.
     */
    public HazardEvent() {

    }

    /**
     * Construct a copy instance.
     * 
     * @param event
     *            Hazard event to be copied.
     */
    public HazardEvent(IReadableHazardEvent event) {
        this();
        setModified(event.isModified());
        setEventID(event.getEventID());
        setSiteID(event.getSiteID());
        setEndTime(event.getEndTime());
        setStartTime(event.getStartTime());
        setCreationTime(event.getCreationTime());
        setInsertTime(event.getInsertTime());
        setGeometry(event.getGeometry());
        setVisualFeatures(event.getVisualFeatures());
        setPhenomenon(event.getPhenomenon());
        setSignificance(event.getSignificance());
        setSubType(event.getSubType());
        setStatus(event.getStatus());
        setHazardMode(event.getHazardMode());
        setSource(event.getSource());
        setWsId(event.getWsId());
        if (event.getHazardAttributes() != null) {
            setHazardAttributes(new HashMap<>(event.getHazardAttributes()));
        }
    }

    @Override
    public boolean isModified() {
        return modified;
    }

    @Override
    public void setModified(boolean modified) {
        this.modified = modified;
    }

    @Override
    public String getSiteID() {
        return siteID;
    }

    @Override
    public void setSiteID(String site) {
        this.siteID = site;
    }

    @Override
    public String getEventID() {
        return eventID;
    }

    @Override
    public void setEventID(String eventId) {
        this.eventID = eventId;
    }

    @Override
    public String getDisplayEventID() {
        return (HazardServicesEventIdUtil.getDisplayId(getEventID()));
    }

    public String getUniqueID() {
        return uniqueID;
    }

    public void setUniqueID(String uniqueID) {
        this.uniqueID = uniqueID;
    }

    public boolean isLatestVersion() {
        return uniqueID.equals(LATEST_VERSION);
    }

    public void setLatestVersion() {
        uniqueID = LATEST_VERSION;
    }

    @Override
    public HazardStatus getStatus() {
        return status;
    }

    @Override
    public void setStatus(HazardStatus state) {
        this.status = state;
    }

    @Override
    public String getPhenomenon() {
        return phenomenon;
    }

    @Override
    public void setPhenomenon(String phenomenon) {
        this.phenomenon = phenomenon;
    }

    @Override
    public String getSignificance() {
        return significance;
    }

    @Override
    public void setSignificance(String significance) {
        this.significance = significance;
    }

    @Override
    public String getSubType() {
        return subType;
    }

    @Override
    public void setSubType(String subType) {
        this.subType = subType;
    }

    @Override
    public String getHazardType() {
        return HazardEventUtilities.getHazardType(this);
    }

    @Override
    public void setHazardType(String phenomenon, String significance,
            String subtype) {
        setPhenomenon(phenomenon);
        setSignificance(significance);
        setSubType(subtype);
    }

    @Override
    public String getPhensig() {
        return HazardEventUtilities.getHazardPhenSig(this);
    }

    @Override
    public Date getStartTime() {
        return startTime;
    }

    @Override
    public void setStartTime(Date startTime) {
        this.startTime = new Date(startTime.getTime());
    }

    @Override
    public Date getEndTime() {
        return endTime;
    }

    @Override
    public void setEndTime(Date endTime) {
        this.endTime = new Date(endTime.getTime());
    }

    @Override
    public void setTimeRange(Date startTime, Date endTime) {
        setStartTime(startTime);
        setEndTime(endTime);
    }

    @Override
    public Date getCreationTime() {
        return creationTime;
    }

    @Override
    public void setCreationTime(Date creationTime) {
        this.creationTime = new Date(creationTime.getTime());
    }

    @Override
    public ProductClass getHazardMode() {
        return hazardMode;
    }

    @Override
    public void setHazardMode(ProductClass hazardMode) {
        this.hazardMode = hazardMode;
    }

    @Override
    public Source getSource() {
        return source;
    }

    @Override
    public void setSource(Source source) {
        this.source = source;
    }

    @Override
    public Geometry getFlattenedGeometry() {
        return flattenedGeometry;
    }

    @Override
    public IAdvancedGeometry getGeometry() {
        return geometry;
    }

    /**
     * Set the flattened geometry to that specified. Needed for Thrift
     * deserialization.
     * 
     * @param flattenedGeometry
     *            Flattened geometry to be used.
     */
    public void setFlattenedGeometry(Geometry flattenedGeometry) {
        this.flattenedGeometry = flattenedGeometry;
    }

    @Override
    public void setGeometry(IAdvancedGeometry geometry) {
        this.geometry = geometry;
        this.flattenedGeometry = AdvancedGeometryUtilities
                .getJtsGeometryAsCollection(geometry);
    }

    @Override
    public VisualFeature getVisualFeature(String identifier) {
        return (visualFeatures == null ? null
                : visualFeatures.getByIdentifier(identifier));
    }

    @Override
    public VisualFeaturesList getVisualFeatures() {
        return visualFeatures;
    }

    @Override
    public boolean setVisualFeature(VisualFeature visualFeature) {
        if (visualFeatures == null) {
            return false;
        }
        return visualFeatures.replace(visualFeature);
    }

    @Override
    public void setVisualFeatures(VisualFeaturesList visualFeatures) {
        this.visualFeatures = visualFeatures;
    }

    @Override
    public Map<String, Serializable> getHazardAttributes() {
        Map<String, Serializable> attrs = new HashMap<>();

        for (HazardAttribute attribute : attributes) {
            attrs.put(attribute.getKey(),
                    (Serializable) attribute.getValueObject());
        }
        return attrs;
    }

    @Override
    public void addHazardAttribute(final String key, final Serializable value) {

        /*
         * Remove the old hazard attribute first, since given that the
         * attributes are a set of HazardAttribute objects, and HazardAttribute
         * does not implement equals() and hashCode() to only consider the key
         * of the object, adding one here will probably cause two hazard
         * attributes with the same key but different values to be left in the
         * set otherwise.
         */
        removeHazardAttribute(key);
        this.attributes.add(new HazardAttribute(eventID, key, value));

    }

    @Override
    public void addHazardAttributes(Map<String, Serializable> attributes) {
        for (Map.Entry<String, Serializable> entry : attributes.entrySet()) {
            addHazardAttribute(entry.getKey(), entry.getValue());
        }
    }

    @Override
    public Serializable getHazardAttribute(String key) {
        return getHazardAttributes().get(key);
    }

    @Override
    public void removeHazardAttribute(String key) {
        HazardAttribute attributeToRemove = null;
        for (HazardAttribute attribute : attributes) {
            if (attribute.getKey().equals(key)) {
                attributeToRemove = attribute;
                break;
            }
        }
        if (attributeToRemove != null) {
            attributes.remove(attributeToRemove);
        }
    }

    @Override
    public boolean isValid() throws ValidationException {
        /*
         * TODO: Determine validation strategy.
         */
        return true;
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("EventId : ").append(eventID).append("\n");
        builder.append("Site : ").append(siteID).append("\n");
        builder.append("Phensig : ").append(phenomenon).append(".")
                .append(significance).append("\n");
        builder.append("Creation Time : ")
                .append(new Date(creationTime.getTime())).append("\n");
        builder.append("Start Time : ").append(new Date(startTime.getTime()))
                .append("\n");
        builder.append("End Time : ").append(new Date(endTime.getTime()))
                .append("\n");
        if (this.attributes.isEmpty() == false) {
            builder.append("--Attributes--\n");
            for (HazardAttribute attr : this.attributes) {
                builder.append(attr.toString());
            }
        }
        return builder.toString();
    }

    @Override
    public int hashCode() {
        final int prime = 31;
        int result = 1;
        result = prime * result + (modified ? 1 : 0);
        result = prime * result
                + ((creationTime == null) ? 0 : creationTime.hashCode());
        result = prime * result + ((endTime == null) ? 0 : endTime.hashCode());
        result = prime * result + ((eventID == null) ? 0 : eventID.hashCode());
        result = prime * result
                + ((geometry == null) ? 0 : geometry.hashCode());
        result = prime * result
                + ((visualFeatures == null) ? 0 : visualFeatures.hashCode());
        result = prime * result
                + ((attributes == null) ? 0 : attributes.hashCode());
        result = prime * result
                + ((hazardMode == null) ? 0 : hazardMode.hashCode());
        result = prime * result + ((source == null) ? 0 : source.hashCode());
        result = prime * result
                + ((insertTime == null) ? 0 : insertTime.hashCode());
        result = prime * result
                + ((phenomenon == null) ? 0 : phenomenon.hashCode());
        result = prime * result
                + ((significance == null) ? 0 : significance.hashCode());
        result = prime * result + ((siteID == null) ? 0 : siteID.hashCode());
        result = prime * result
                + ((startTime == null) ? 0 : startTime.hashCode());
        result = prime * result + ((status == null) ? 0 : status.hashCode());
        result = prime * result + ((subType == null) ? 0 : subType.hashCode());
        result = prime * result
                + ((uniqueID == null) ? 0 : uniqueID.hashCode());
        return result;
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) {
            return true;
        }
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        HazardEvent other = (HazardEvent) obj;
        if (uniqueID == null) {
            if (other.uniqueID != null) {
                return false;
            }
        } else if (!uniqueID.equals(other.uniqueID)) {
            return false;
        }
        if (creationTime == null) {
            if (other.creationTime != null) {
                return false;
            }
        } else if (!creationTime.equals(other.creationTime)) {
            return false;
        }
        if (endTime == null) {
            if (other.endTime != null) {
                return false;
            }
        } else if (!endTime.equals(other.endTime)) {
            return false;
        }
        if (eventID == null) {
            if (other.eventID != null) {
                return false;
            }
        } else if (!eventID.equals(other.eventID)) {
            return false;
        }
        if (geometry == null) {
            if (other.geometry != null) {
                return false;
            }
        } else if (!geometry.equals(other.geometry)) {
            return false;
        }
        if (visualFeatures == null) {
            if (other.visualFeatures != null) {
                return false;
            }
        } else if (!visualFeatures.equals(other.visualFeatures)) {
            return false;
        }
        if (attributes == null) {
            if (other.attributes != null) {
                return false;
            }
        } else if (!attributes.equals(other.attributes)) {
            return false;
        }
        if (hazardMode != other.hazardMode) {
            return false;
        }
        if (source != other.source) {
            return false;
        }
        if (insertTime == null) {
            if (other.insertTime != null) {
                return false;
            }
        } else if (!insertTime.equals(other.insertTime)) {
            return false;
        }
        if (phenomenon == null) {
            if (other.phenomenon != null) {
                return false;
            }
        } else if (!phenomenon.equals(other.phenomenon)) {
            return false;
        }
        if (significance == null) {
            if (other.significance != null) {
                return false;
            }
        } else if (!significance.equals(other.significance)) {
            return false;
        }
        if (siteID == null) {
            if (other.siteID != null) {
                return false;
            }
        } else if (!siteID.equals(other.siteID)) {
            return false;
        }
        if (startTime == null) {
            if (other.startTime != null) {
                return false;
            }
        } else if (!startTime.equals(other.startTime)) {
            return false;
        }
        if (status != other.status) {
            return false;
        }
        if (subType == null) {
            if (other.subType != null) {
                return false;
            }
        } else if (!subType.equals(other.subType)) {
            return false;
        }
        return (modified == other.modified);
    }

    @Override
    public void setInsertTime(Date date) {
        this.insertTime = date;

    }

    @Override
    public Date getInsertTime() {
        return this.insertTime;
    }

    @Override
    public void setHazardAttributes(Map<String, Serializable> attributes) {
        this.attributes.clear();
        for (Entry<String, Serializable> entry : attributes.entrySet()) {
            addHazardAttribute(entry.getKey(), entry.getValue());
        }
    }

    public Set<HazardAttribute> getAttributes() {
        return attributes;
    }

    public void setAttributes(Set<HazardAttribute> attributes) {
        this.attributes = attributes;
    }

    @Override
    public Geometry getProductGeometry() {
        return HazardEventUtilities.getProductGeometry(this);
    }

    @Override
    public void setProductGeometry(Geometry geom) {
        HazardEventUtilities.setProductGeometry(this, geom);
    }

    @Override
    public WsId getWsId() {
        return wsId;
    }

    @Override
    public void setWsId(WsId wsId) {
        this.wsId = wsId;
    }
}
